features/steps/steps.py

```diff
@@ -1,33 +1,294 @@
+    #return o_doi_label
+
+def draft_output_compare(output_file, ref_output_value):
+    # Function compare two XML files created from 'draft' or 'reserve' actions.
+    # Assumption(s): 
+    #   1.  The name of the new XML file is defined in get_temporary_output_filename().
+    #   2.  The name of the historical name is ref_output_value
+    logger.info(f"output_file,ref_output_value {output_file},{ref_output_value}")
+
+    o_fields_differ_list, o_values_differ_list, o_record_index_differ_list = DOIDiffer.doi_xml_differ(ref_output_value,
+                                                                                                      output_file)
+
+
+    logger.info(f'different fields are {o_fields_differ_list}')
+
+    assert len(o_fields_differ_list) is 0
+
+    return 1
 
-@given('a valid PDS4 {input_type} at url {input_value}')
-def given_valid_input(context, input_type, input_value):
-    logger.info(f"given {input_type} {input_value}")
+def reserve_output_compare(output_file, ref_output_value):
+    logger.info(f"output_file,ref_output_value {output_file},{ref_output_value}")
+    # Use the same function draft_output_compare() to compare a DOI (XML output) from the 'reserve' action.
+    draft_output_compare(output_file,ref_output_value)
 
+def initialize_normal_draft():
+    global g_action_expectant_bad, g_draft_condition
+    # Set to g_action_expectant_bad to signify to not expect the action to be bad.
+    g_action_expectant_bad = False
+    g_draft_condition = TestCondition.NORMAL
 
-@given('an invalid PDS4 {input_type} at {input_value}')
+def initialize_normal_reserve():
+    global g_action_expectant_bad, g_reserve_condition
+    # Set to g_action_expectant_bad to signify to not expect the action to be bad.
+    g_action_expectant_bad = False
+    g_reserve_condition = TestCondition.NORMAL
+
+@given('a valid PDS4 label at {input_value}')
+def given_valid_draft_input(context, input_value):
+    global g_action_expectant_bad, g_draft_condition
+    logger.info(f"given {input_value}")
+    initialize_normal_draft()
+    context.input_value = input_value
+    logger.info(f"g_action_expectant_bad {g_action_expectant_bad}")
+
+@given('a valid reserve PDS4 label at input_type,input_value {input_type},{input_value}')
+def given_valid_reserve_input(context, input_type, input_value):
+    global g_action_expectant_bad, g_reserve_condition
+    logger.info(f"given {input_type},{input_value}")
+    initialize_normal_reserve()
+    context.input_type  = input_type
+    context.input_value = input_value
+    logger.info(f"g_action_expectant_bad {g_action_expectant_bad}")
+
+@given('an invalid PDS4 label at input_type,input_value {input_type},{input_value}')
 def given_invalid_pds4(context, input_type, input_value):
-    raise NotImplementedError(
-        u'STEP: an invalid PDS4 {input_type} at {input_value}')
+    logger.info(f'an invalid PDS4 label at input_type,input_value {input_type},{input_value}')
```

Comment:
The handling of exception is done in function that deal with drafting an DOI from PDS4 label.  The function given_invalid_pds4 currently serve to have something to corresponding to the snippet in feature file to avoid having error with unimplemented code.

Reply:
ok, note that here I have added the initialization of the input_value in the context